From 76f1ce4a01b1d060dee30fc0e21ae99db60bab7c Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Fri, 5 Sep 2014 16:36:16 -0400
Subject: [target] Fix alua_access_state attribute OOPs for un-configured devices

Message-id: <1409934989-29941-19-git-send-email-agrover@redhat.com>
Patchwork-id: 91481
O-Subject: [RHEL7.1 PATCH e-stor 18/31] target: Fix alua_access_state attribute OOPs for un-configured devices
Bugzilla: 1129387
RH-Acked-by: Chris Leech <cleech@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

From: Nicholas Bellinger <nab@linux-iscsi.org>

commit f1453773514bb8b0bba0716301e8c8f17f8d39c7 upstream.

This patch fixes a OOPs where an attempt to write to the per-device
alua_access_state configfs attribute at:

  /sys/kernel/config/target/core/$HBA/$DEV/alua/$TG_PT_GP/alua_access_state

results in an NULL pointer dereference when the backend device has not
yet been configured.

This patch adds an explicit check for DF_CONFIGURED, and fails with
-ENODEV to avoid this case.

Reported-by: Chris Boot <crb@tiger-computing.co.uk>
Reported-by: Philip Gaw <pgaw@darktech.org.uk>
Cc: Chris Boot <crb@tiger-computing.co.uk>
Cc: Philip Gaw <pgaw@darktech.org.uk>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/target/target_core_configfs.c b/drivers/target/target_core_configfs.c
index 4a8bd36..8cda408 100644
--- a/drivers/target/target_core_configfs.c
+++ b/drivers/target/target_core_configfs.c
@@ -2034,6 +2034,11 @@ static ssize_t target_core_alua_tg_pt_gp_store_attr_alua_access_state(
    " tg_pt_gp ID: %hu\n", tg_pt_gp->tg_pt_gp_valid_id);
   return -EINVAL;
  }
+ if (!(dev->dev_flags & DF_CONFIGURED)) {
+  pr_err("Unable to set alua_access_state while device is"
+         " not configured\n");
+  return -ENODEV;
+ }
 
  ret = strict_strtoul(page, 0, &tmp);
  if (ret < 0) {
-- 
1.7.1